File: apps/backend-gateway/src/interfaces/middleware/rateLimiter.ts
/**
 * StadiumOS AI — Express Rate Limiting Middleware.
 * 
 * Configures express-rate-limit utilizing Redis as distributed store cache,
 * with standard in-memory store fallback.
 */

import rateLimit from "express-rate-limit";
import RedisStore from "rate-limit-redis";
import { config } from "../../config";
import { getRedisClient } from "../../infrastructure/cache/redisClient";
import { logger } from "../../infrastructure/logging/logger";

let redisAvailable = false;

try {
  getRedisClient();
  redisAvailable = true;
} catch (err: any) {
  logger.warn("Redis client uninitialized. Falling back to local in-memory rate-limit store:", err.message);
}

let storeCounter = 0;

function createStore() {
  if (!redisAvailable) return undefined;
  try {
    const redis = getRedisClient();
    storeCounter++;
    return new RedisStore({
      // @ts-ignore
      sendCommand: (...args: string[]) => redis.call(args[0], ...args.slice(1)),
      prefix: `rl:${storeCounter}:`
    });
  } catch {
    return undefined;
  }
}

export function createRateLimiter(limitCount: number = config.rateLimitPerMinute, windowMs: number = 60 * 1000) {
  return rateLimit({
    windowMs,
    max: limitCount,
    standardHeaders: true,
    legacyHeaders: false,
    store: createStore(),
    handler: (req, res) => {
      logger.warn({ ip: req.ip, path: req.url }, "Rate limit threshold breached");
      res.status(429).json({
        success: false,
        error_code: "RATE_LIMIT_EXCEEDED",
        message: "Too many requests. Please slow down and try again later."
      });
    }
  });
}

// Default system rate limiter
export const defaultRateLimiter = createRateLimiter();

File: apps/backend-gateway/src/interfaces/middleware/rate_limiter.py
"""
StadiumOS AI — Rate Limiting Configuration.

Initializes the slowapi rate limiter using the remote address key generator
and links it to Redis if available.
"""

from fastapi import Request
from slowapi import Limiter
from slowapi.util import get_remote_address
from slowapi.errors import RateLimitExceeded
from fastapi.responses import JSONResponse
import structlog

from src.config import get_settings

logger = structlog.get_logger("rate_limiter")

settings = get_settings()

# Initialize the global rate limiter using the client's IP address.
# In production, slowapi connects to the Redis server for distributed tracking.
limiter = Limiter(
    key_func=get_remote_address,
    storage_uri=settings.redis_url,
    default_limits=[f"{settings.rate_limit_per_minute}/minute"],
)


async def rate_limit_exceeded_handler(request: Request, exc: RateLimitExceeded) -> JSONResponse:
    """
    Handle rate limit exceeded exceptions.

    Formats the response to match the unified system error schema (429 status code).
    """
    client_ip = get_remote_address(request)
    await logger.awarn(
        "rate_limit_exceeded",
        client_ip=client_ip,
        path=request.url.path,
        limit=exc.detail,
    )
    return JSONResponse(
        status_code=429,
        content={
            "success": False,
            "error_code": "RATE_LIMIT_EXCEEDED",
            "message": "Too many requests. Please slow down and try again later.",
            "retry_after": exc.detail,
        },
        headers={"X-RateLimit-Limit": exc.detail},
    )
